Finish "Made with Dioxus", add "Our projects", etc., refactor code

tigerros opened the pull request:
New features:
- Finished the "Made with Dioxus" page ([screenshot](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/21cdabbf-e6d7-4b6f-8611-93ef3534fcb5)) with some projects.
- Added an "Our projects" page ([screenshot](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/0b1cc098-fd23-4565-9b2a-8a477b7252fc)). This is an automatically generated list of projects from the organization. It looks for a `Cargo.toml` file at the root of each repository to determine whether it should be included. However, I think a file like `.dxcom-project` would be better since it's more explicit, and not all projects might have a `Cargo.toml` at the root.
- Added `build.rs` to run Tailwind automatically on each build. Works with hot reload. Also installed Tailwind in the GitHub workflow (which should be ran by `build.rs`), hopefully it works in there too.
- Updated the README to reflect all changes.

Refactors:
- Added two new directories to `src`; `models` and `components`. Models contain simple structs/enums like `Project` and `ProjectCategory`. Components contain components that are used by the routes.
- Added `src/projects.rs` to make adding projects easier for people who want their project there.
- Added a layout (`components/layout.rs`) to the routes to include the navigation everywhere, and set size and position constraints for the `Outlet::<Route>`.
- Changed the rounding of the navigation to be consistent with the project cards (Tailwind `rounded-md`). Also looks better on mobile IMO.
- Fix some layout issues on mobile.

Small refactors based on my opinion:
- Renamed `NavBar` component to `Navigation`. I think it's cleaner without an acronym, and it might not be a "bar" forever.
- Moved `input.css` to `src/input.css`. I think it just makes more sense.

New dependencies:
- [`dry-mods`](https://crates.io/crates/dry-mods) - Simple macro I made to reduce the boilerplate in `mod.rs` files and preludes. Currently using it to export the contents of modules in `components`, `models` and `routes`.
- `reqwest` - For fetching GitHub stars.

marc2332 reviewed (changes requested):
I think a 2 fixed-size columns would be better here 😄 

![image](https://github.com/dioxus-community/dioxus-community.github.io/assets/38158676/07824058-3bbe-4ffd-af60-768496d02fef)

marc2332 commented on src/components/navigation.rs:
```diff
@@ -0,0 +1,18 @@
+        nav { class: "mx-auto p-2 flex justify-stretch bg-blue-1 rounded-md gap-2",
+            Link { class: NAVBAR_LINK_STYLE, to: Route::Home {}, "Home" }
+            Link { class: NAVBAR_LINK_STYLE, to: "https://github.com/dioxus-community", "GitHub" }
+            Link { class: NAVBAR_LINK_STYLE, to: Route::MadeWithDioxus {}, "Made with Dioxus" }
+            Link { class: NAVBAR_LINK_STYLE, to: Route::OurProjects {}, "Our projects" }
```
swap places
```suggestion
            Link { class: NAVBAR_LINK_STYLE, to: Route::OurProjects {}, "Our projects" }
            Link { class: NAVBAR_LINK_STYLE, to: Route::MadeWithDioxus {}, "Made with Dioxus" }
```

marc2332 commented on src/components/navigation.rs:
```diff
@@ -0,0 +1,18 @@
+#[inline_props]
+pub fn Navigation(cx: Scope) -> Element {
+    render! {
+        nav { class: "mx-auto p-2 flex justify-stretch bg-blue-1 rounded-md gap-2",
```
I think a fully rounded navbar is better 😝

tigerros replied:
I don't know... I kind of like the consistency with the project cards. It also becomes a circle on small screens :/

tigerros replied:
It now uses the fully rounded version on medium screens and wider, but uses the medium version on smaller screens. Trust me, the fully rounded version looks really bad on small screens.

marc2332 commented:
> Added an "Our projects" page ([screenshot](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/0cc0dd63-73f2-4c65-a658-4ed4d065ea1b)). This is an automatically generated list of projects from the organization. It looks for a Cargo.toml file at the root of each repository to determine whether it should be included. However, I think a file like .dxcom-project would be better since it's more explicit, and not all projects might have a Cargo.toml at the root.

I also think it would be better to have a static file with the projects, it will let us be more flexible. With the cost of not showing the stars though, which I think it's fine anyway.

tigerros commented:
> I think a 2 fixed-size columns would be better for "Made with Dioxus" and "Our Projects"😄

![image](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/1100c5d0-b23e-4ba1-b85a-e2b7adf59a8e)

Looks a little empty, but it's really neat! Also easier to style :P

marc2332 commented on src/projects.rs:
Some of these look great, but I think it would be better to remove `outdated/very basic/without demo` projects.

I think we could add https://github.com/marc2332/freya-editor for example too

tigerros replied:
Yeah the list isn't that high effort. To be fair, I couldn't find that many projects that weren't libraries. I wanted to ask other people what should be there.

There's also something else I'm concerned about... Awesome Dioxus. It's pretty similar. In fact, awesome lists often have real world apps too. It really feels like Awesome Dioxus and Made with Dioxus should be in the same place. Maybe this should instead be in the official docs? Then at the same time, an awesome list makes more sense to be maintained by the community. I guess it sorts of depends on what this organization is aiming to be.

marc2332 replied:
I wasn't really sure about doing "Made with Dioxus" yet, hence why I left it commented 😅 Perhaps it would just be better to keep up to date the Awesome list from dioxus labs

tigerros replied:
We could ask @ealmloff or @jkelleyrtp

ealmloff replied:
> We could ask @ealmloff or @jkelleyrtp

Adding it to the main Dioxus awsome page would be great!

tigerros replied:
> Adding it to the main Dioxus awsome page would be great!

Sounds good

tigerros commented:
> > Added an "Our projects" page ([screenshot](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/0cc0dd63-73f2-4c65-a658-4ed4d065ea1b)). This is an automatically generated list of projects from the organization. It looks for a Cargo.toml file at the root of each repository to determine whether it should be included. However, I think a file like .dxcom-project would be better since it's more explicit, and not all projects might have a Cargo.toml at the root.
> 
> I also think it would be better to have a static file with the projects (like it does with "Made with Dioxus"), it will let us be more flexible. With the cost of not showing the stars though, which I think it's fine anyway.

I mean, we *could* get the stars. We could fetch them in the `ProjectCard` component. However, I also had the idea of making a proc macro to get the stars at compile time, but that might be overkill. It also wouldn't be real-time, but I think we can schedule some automatic building to keep the stars sort of accurate. Also might be overkill though.

marc2332 commented:
> > > Added an "Our projects" page ([screenshot](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/0cc0dd63-73f2-4c65-a658-4ed4d065ea1b)). This is an automatically generated list of projects from the organization. It looks for a Cargo.toml file at the root of each repository to determine whether it should be included. However, I think a file like .dxcom-project would be better since it's more explicit, and not all projects might have a Cargo.toml at the root.
> > 
> > 
> > I also think it would be better to have a static file with the projects (like it does with "Made with Dioxus"), it will let us be more flexible. With the cost of not showing the stars though, which I think it's fine anyway.
> 
> I mean, we _could_ get the stars. We could fetch them in the `ProjectCard` component. However, I also had the idea of making a proc macro to get the stars at compile time, but that might be overkill. It also wouldn't be real-time, but I think we can schedule some automatic building to keep the stars sort of accurate. Also might be overkill though.

That definitely sounds overkill, I think a static approach is the way to go

tigerros commented:
> > > > Added an "Our projects" page ([screenshot](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/0cc0dd63-73f2-4c65-a658-4ed4d065ea1b)). This is an automatically generated list of projects from the organization. It looks for a Cargo.toml file at the root of each repository to determine whether it should be included. However, I think a file like .dxcom-project would be better since it's more explicit, and not all projects might have a Cargo.toml at the root.
> > > 
> > > 
> > > I also think it would be better to have a static file with the projects (like it does with "Made with Dioxus"), it will let us be more flexible. With the cost of not showing the stars though, which I think it's fine anyway.
> > 
> > 
> > I mean, we _could_ get the stars. We could fetch them in the `ProjectCard` component. However, I also had the idea of making a proc macro to get the stars at compile time, but that might be overkill. It also wouldn't be real-time, but I think we can schedule some automatic building to keep the stars sort of accurate. Also might be overkill though.
> 
> That definitely sounds overkill, I think a static approach is the way to go

Ok, I'll add it.

marc2332 commented:
> > I think a 2 fixed-size columns would be better for "Made with Dioxus" and "Our Projects"😄
> 
> ![image](https://user-images.githubusercontent.com/76173380/274716401-1100c5d0-b23e-4ba1-b85a-e2b7adf59a8e.png)
> 
> Looks a little empty, but it's really neat! Also easier to style :P

Nice, maybe try reducing the width of the cards, so they are more centered ✨

tigerros commented:
> Nice, maybe try reducing the width of the cards, so they are more centered ✨

What about this? 
![image](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/260d1a7e-0792-45c5-b969-a41d9ca8eacd)

tigerros commented:
Added the const for our projects and also added categories. They're not displayed though, since there's only two types being used now: "App" and "Utility". I'll probably create some more categories for state, hooks, etc. Any other ideas?

marc2332 commented:
> > Nice, maybe try reducing the width of the cards, so they are more centered ✨
> 
> What about this?
> 
> ![image](https://user-images.githubusercontent.com/76173380/274723718-260d1a7e-0792-45c5-b969-a41d9ca8eacd.png)

Maybe even less width and I think it will look great !

tigerros commented:
> Maybe even less width and I think it will look great !

Doubled it:

![image](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/da83adf9-c09b-4a40-9260-d22fb36238c5)

tigerros commented:
I was using side padding. Not sure what I was thinking! Completely unusable on mobile. Current style:

![image](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/2421e552-f5a7-4ee8-a097-39c6b9af1402)

Mobile:

![image](https://github.com/dioxus-community/dioxus-community.github.io/assets/76173380/80873b19-4fe8-4ac9-9886-766bcdf2fa92)

tigerros commented:
@marc2332 Some new stuff:
- Active style for the navigation routes.
- Fetching stars at runtime (a `insert_stars` boolean property in the `ProjectCard` component).

I think it's ready to be merged. Nothing else to improve comes to mind.

marc2332 commented on src/components/project_card.rs:
```diff
@@ -0,0 +1,155 @@
+
+            None
+        },
+        None => None,
```
Could we maybe still show it as `⭐ Stars            ...` instead of being empty ? I think it will be better as it won't cause the layout to shift

tigerros replied:
Yeah probably better that way.

marc2332 reviewed (changes requested):
Just one small thing! Looks very good ⚡

srid commented:
> * Added `build.rs` to run Tailwind automatically on each build. Works with hot reload. Also installed Tailwind in the GitHub workflow (which should be ran by `build.rs`), hopefully it works in there too.

Is this something that can be upstreamed to Dioxus so others projects using tailwind can benefit from it automatically?

tigerros commented:
> Is this something that can be upstreamed to Dioxus so others projects using tailwind can benefit from it automatically?

Once the CLI plugin system is finished, this will likely be the first plugin ;) Right now, I don't think there's a point in documenting this trivial build script.

marc2332 commented:
Amazing work @tigerros 🔥
